File: dqops/src/main/java/com/dqops/data/checkresults/services/models/CheckResultsListModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.data.checkresults.services.models;

import com.dqops.checks.CheckType;
import com.dqops.rest.models.metadata.RuleListModel;
import com.dqops.utils.docs.generators.SampleListUtility;
import com.dqops.utils.docs.generators.SampleLongsRegistry;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.Data;

import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalAccessor;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Check detailed results. Returned in the context of a single data group, with a supplied list of other data groups.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.LowerCamelCaseStrategy.class)
@Data
public class CheckResultsListModel {
    @JsonPropertyDescription("Check hash")
    private long checkHash;

    @JsonPropertyDescription("Check category name")
    private String checkCategory;

    @JsonPropertyDescription("Check name")
    private String checkName;

    @JsonPropertyDescription("Check display name")
    private String checkDisplayName;

    @JsonPropertyDescription("Check type")
    private CheckType checkType;

    @JsonPropertyDescription("Data groups list")
    private List<String> dataGroups;

    @JsonPropertyDescription("Selected data group")
    private String dataGroup;

    @JsonPropertyDescription("Single check results")
    private List<CheckResultEntryModel> checkResultEntries = new ArrayList<>();

    public static class CheckResultsListModelSampleFactory implements SampleValueFactory<CheckResultsListModel> {
        @Override
        public CheckResultsListModel createSample() {
            List<CheckResultEntryModel> checkResultEntryModels = SampleListUtility.generateList(
                    CheckResultEntryModel.class, 4,

                    CheckResultEntryModel::getId,
                    s -> Long.toString(new Random(Long.parseLong(s)).nextLong()),
                    CheckResultEntryModel::setId,

                    CheckResultEntryModel::getExecutedAt,
                    i -> i.plus(1L, ChronoUnit.HOURS),
                    CheckResultEntryModel::setExecutedAt,

                    CheckResultEntryModel::getTimePeriod,
                    d -> d.plusHours(1L),
                    CheckResultEntryModel::setTimePeriod
            );
            for (int i = 0; i < checkResultEntryModels.size(); ++i) {
                CheckResultEntryModel c = checkResultEntryModels.get(i);
                c.setActualValue(c.getActualValue() + 10*i);
                c.setExpectedValue(c.getExpectedValue() + 10*i);
                c.setWarningLowerBound(c.getWarningLowerBound() + 10*i);
                c.setWarningUpperBound(c.getWarningUpperBound() + 10*i);
                c.setErrorLowerBound(c.getErrorLowerBound() + 10*i);
                c.setErrorUpperBound(c.getErrorUpperBound() + 10*i);
                c.setFatalLowerBound(c.getFatalLowerBound() + 10*i);
                c.setFatalUpperBound(c.getFatalUpperBound() + 10*i);
            }

            CheckResultEntryModel c = checkResultEntryModels.get(0);

            return new CheckResultsListModel() {{
                setCheckHash(c.getCheckHash());
                setCheckCategory(c.getCheckCategory());
                setCheckName(c.getCheckName());
                setCheckDisplayName(c.getCheckDisplayName());
                setCheckType(c.getCheckType());
                setDataGroup(c.getDataGroup());
                setCheckResultEntries(checkResultEntryModels);
            }};
        }
    }
}

File: distribution/python/dqops/client/models/check_results_list_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union, cast

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.check_type import CheckType
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.check_result_entry_model import CheckResultEntryModel


T = TypeVar("T", bound="CheckResultsListModel")


@_attrs_define
class CheckResultsListModel:
    """
    Attributes:
        check_hash (Union[Unset, int]): Check hash
        check_category (Union[Unset, str]): Check category name
        check_name (Union[Unset, str]): Check name
        check_display_name (Union[Unset, str]): Check display name
        check_type (Union[Unset, CheckType]):
        data_groups (Union[Unset, List[str]]): Data groups list
        data_group (Union[Unset, str]): Selected data group
        check_result_entries (Union[Unset, List['CheckResultEntryModel']]): Single check results
    """

    check_hash: Union[Unset, int] = UNSET
    check_category: Union[Unset, str] = UNSET
    check_name: Union[Unset, str] = UNSET
    check_display_name: Union[Unset, str] = UNSET
    check_type: Union[Unset, CheckType] = UNSET
    data_groups: Union[Unset, List[str]] = UNSET
    data_group: Union[Unset, str] = UNSET
    check_result_entries: Union[Unset, List["CheckResultEntryModel"]] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        check_hash = self.check_hash
        check_category = self.check_category
        check_name = self.check_name
        check_display_name = self.check_display_name
        check_type: Union[Unset, str] = UNSET
        if not isinstance(self.check_type, Unset):
            check_type = self.check_type.value

        data_groups: Union[Unset, List[str]] = UNSET
        if not isinstance(self.data_groups, Unset):
            data_groups = self.data_groups

        data_group = self.data_group
        check_result_entries: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.check_result_entries, Unset):
            check_result_entries = []
            for check_result_entries_item_data in self.check_result_entries:
                check_result_entries_item = check_result_entries_item_data.to_dict()

                check_result_entries.append(check_result_entries_item)

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if check_hash is not UNSET:
            field_dict["checkHash"] = check_hash
        if check_category is not UNSET:
            field_dict["checkCategory"] = check_category
        if check_name is not UNSET:
            field_dict["checkName"] = check_name
        if check_display_name is not UNSET:
            field_dict["checkDisplayName"] = check_display_name
        if check_type is not UNSET:
            field_dict["checkType"] = check_type
        if data_groups is not UNSET:
            field_dict["dataGroups"] = data_groups
        if data_group is not UNSET:
            field_dict["dataGroup"] = data_group
        if check_result_entries is not UNSET:
            field_dict["checkResultEntries"] = check_result_entries

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.check_result_entry_model import CheckResultEntryModel

        d = src_dict.copy()
        check_hash = d.pop("checkHash", UNSET)

        check_category = d.pop("checkCategory", UNSET)

        check_name = d.pop("checkName", UNSET)

        check_display_name = d.pop("checkDisplayName", UNSET)

        _check_type = d.pop("checkType", UNSET)
        check_type: Union[Unset, CheckType]
        if isinstance(_check_type, Unset):
            check_type = UNSET
        else:
            check_type = CheckType(_check_type)

        data_groups = cast(List[str], d.pop("dataGroups", UNSET))

        data_group = d.pop("dataGroup", UNSET)

        check_result_entries = []
        _check_result_entries = d.pop("checkResultEntries", UNSET)
        for check_result_entries_item_data in _check_result_entries or []:
            check_result_entries_item = CheckResultEntryModel.from_dict(
                check_result_entries_item_data
            )

            check_result_entries.append(check_result_entries_item)

        check_results_list_model = cls(
            check_hash=check_hash,
            check_category=check_category,
            check_name=check_name,
            check_display_name=check_display_name,
            check_type=check_type,
            data_groups=data_groups,
            data_group=data_group,
            check_result_entries=check_result_entries,
        )

        check_results_list_model.additional_properties = d
        return check_results_list_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
